### PR TITLE
fix(training)!: expands graph config for multiple datasets

### PR DESCRIPTION
**Pull request by JPXKQX**
## Description

Following discussion around configuration usability for multiple datasets, this PR provides support for different encoder/decoder subgraphs for each dataset. 

This aligns with the work described in ["Training with Observations and gridded datasets"](https://github.com/ecmwf/anemoi-core/milestone/5) Milestone. Once the graph creation is moved deep into the model, we expect to create the different data (global, regional, cerra, ...) nodes directly from the dataset removing them from the graph config.

This breaks the configuration files for the multiple dataset use cases, as we have now to explicitly define the full graph.

This PR tackles issue #30 too.

***As a contributor to the Anemoi framework, please ensure that your changes include unit tests, updates to any affected dependencies and documentation, and have been tested in a parallel setting  (i.e., with multiple GPUs). As a reviewer, you are also responsible for verifying these aspects and requesting changes if they are not adequately addressed. For guidelines about those please refer to https://anemoi.readthedocs.io/en/latest/***

By opening this pull request, I affirm that all authors agree to the [Contributor License Agreement.](https://github.com/ecmwf/codex/blob/main/Legal/contributor_license_agreement.md)

----
📚 Documentation preview 📚: https://anemoi-training--895.org.readthedocs.build/en/895/

----
📚 Documentation preview 📚: https://anemoi-models--895.org.readthedocs.build/en/895/

----
📚 Documentation preview 📚: https://anemoi-graphs--895.org.readthedocs.build/en/895/

### CODE REVIEW

File: models/src/anemoi/models/models/base.py [hunk 1]
```diff
@@ -113,16 +106,13 @@ def _calculate_shapes_and_indices(self, data_indices: dict) -> None:
             self.input_dim_latent[dataset_name] = self._calculate_input_dim_latent(dataset_name)
```
Comment (havardhhaugen) [inline models/src/anemoi/models/models/base.py:124]: Should this be `self.input_dim_latent[dataset_name] = self._calculate_input_dim_latent(self._graph_name_hidden)` ? (Related to the change in `_calculate_input_dim_latent`)

I can see that trainable_parameters.hidden is used for everything, but I think this will break if hidden and data have a different set of node attributes?

File: models/src/anemoi/models/models/autoencoder.py [hunk 2]
```diff
@@ -59,10 +60,10 @@ def _build_networks(self, model_config: DotDict) -> None:
         self.decoder = torch.nn.ModuleDict()
         for dataset_name in self._graph_data.keys():
```
Comment (havardhhaugen) [inline models/src/anemoi/models/models/autoencoder.py]: should this also be `self.dataset_names`?

Comment (JPXKQX) [reply models/src/anemoi/models/models/autoencoder.py]: yes

File: models/src/anemoi/models/models/base.py (hunk 1 above)
Comment (JPXKQX) [reply models/src/anemoi/models/models/base.py:124]: 1) True, updated!
2) Yes, I have updated now. It will raise an error if the trainable parameters are not specified for one of the nodes.

File: models/src/anemoi/models/models/base.py [hunk 3]
```diff
@@ -59,26 +59,19 @@ def __init__(
...
+        self.dataset_names = list(data_indices.keys())
+
         model_config = DotDict(model_config)
-        self._graph_name_data = (
-            model_config.graph.data
-        )  # assumed to be all the same because this is how we construct the graphs
-        self._graph_name_hidden = (
-            model_config.graph.hidden
-        )  # assumed to be all the same because this is how we construct the graphs
+        self._graph_name_hidden = model_config.model.model.hidden_nodes_name
```
Comment (icedoom888) [inline models/src/anemoi/models/models/base.py:66]: Not the biggest fan of having the hidden_nodes written explicitely in the **model** config.

I suggest we do: 
`self._graph_name_hidden=graph_data.keys() - self.dataset_names`
As you do in models/layers/graph.py and filter out the data nodes by name.
This way you only define the hidden nodes names in the graph config (when creating the nodes) and you don't even have to make sure that those names match the model config.

File: models/src/anemoi/models/schemas/models.py [hunk 4]
```diff
@@ -69,6 +69,8 @@ class DefinedModels(str, Enum):
 class Model(BaseModel):
     target_: DefinedModels = Field(..., alias="_target_")
     "Model object defined in anemoi.models.model."
+    hidden_nodes_name: str | list[str] = Field(examples=["hidden", ["hidden1", "hidden2"]])
```
Comment (icedoom888) [inline models/src/anemoi/models/schemas/models.py:71]: Same comment as above: I would not mix graph specific names in the model config.

**Review by icedoom888 — CHANGES_REQUESTED** [review]
LGTM, but I would remove model_config.model.model.hidden_nodes_name.

File: models/src/anemoi/models/models/base.py (hunk 3 above)
Comment (JPXKQX) [reply models/src/anemoi/models/models/base.py:66]: I don't think that would work:
- For hierarchical models, the order is important.
- For ensemble models (or other), there may be extra sets of nodes used for the multi-scale loss or the residual truncation.

File: models/src/anemoi/models/schemas/models.py (hunk 4 above)
Comment (JPXKQX) [reply models/src/anemoi/models/schemas/models.py:71]: Indeed, this relates to some model functionality not to the graph directly. When building a graph, this name doesn't mean anything. If you look at its usage in the code, it's only used in the model class (and some callbacks)

File: models/src/anemoi/models/models/base.py (hunk 3 above)
Comment (icedoom888) [reply models/src/anemoi/models/models/base.py:66]: - Order is preserved: the keys mantain the order of creation of the nodes -> order of declaration in the graph config, which solveds the issue for us
- Regarding extra nodes that are not hidden nodes per se, I would argue those should be declaired in a separate list in graphs: `additional_nodes: ["model_scale", "residual_truncation"]` which can be removed here form the list

File: training/src/anemoi/training/config/dataloader/multi.yaml [hunk 5]
```diff
@@ -41,10 +41,10 @@ grid_indices:
   datasets:
     era5:
       _target_: anemoi.training.data.grid_indices.FullGrid
-      nodes_name: ${graph.data}
+      nodes_name: era5
     cerra:
       _target_: anemoi.training.data.grid_indices.FullGrid  # Could be different
-      nodes_name: ${graph.data}
```
Comment (MeraX) [inline training/src/anemoi/training/config/dataloader/multi.yaml]: Would it be possible (and beneficial) in this context to relate the `nodes_name` to the parent yaml key?

File: training/src/anemoi/training/config/graph/encoder_decoder_only.yaml [hunk 6]
```diff
@@ -1,15 +1,12 @@
...
-data: "data"
-hidden: "hidden"
-
 nodes:
   # Data nodes
   data:
     node_builder:
       _target_: anemoi.graphs.nodes.AnemoiDatasetNodes # options: AnemoiDatasetNodes, NPZFileNodes
-      dataset: "PLACEHOLDER_DATASET_WILL_BE_OVERRIDDEN"  # Always overridden from dataloader config
+      dataset: ${dataloader.training.datasets.data.dataset}
```
Comment (MeraX) [inline training/src/anemoi/training/config/graph/encoder_decoder_only.yaml]: 🙏

**Review by MeraX — COMMENTED** [review]
Thank you, @JPXKQX . I think this makes the code cleaner and more explicit.

File: training/src/anemoi/training/config/dataloader/multi.yaml (hunk 5 above)
Comment (JPXKQX) [reply training/src/anemoi/training/config/dataloader/multi.yaml]: Good point! It shouldn't be required anymore

File: training/src/anemoi/training/config/graph/encoder_decoder_only.yaml (hunk 6 above)
Comment (JPXKQX) [reply training/src/anemoi/training/config/graph/encoder_decoder_only.yaml]: I think the idea is to remove these node builders in line with the Observations milestone, following the work on #848.

File: models/src/anemoi/models/models/base.py (hunk 3 above)
Comment (JPXKQX) [reply models/src/anemoi/models/models/base.py:66]: I would still feel it is safer to keep it, as we have never enforced any ordering in the graph creation

Comment (icedoom888) [reply models/src/anemoi/models/models/base.py:66]: Well this aslo assumes ordering in :  ` self._graph_name_hidden = model_config.model.model.hidden_nodes_name` which i think it's not enforced?

File: models/src/anemoi/models/schemas/models.py (hunk 4 above)
Comment (icedoom888) [reply models/src/anemoi/models/schemas/models.py:71]: I still find it confusing since information is shared across very distant parts of the config.
Can we try to polish this here? And add some checks for the naming and the order?

File: models/src/anemoi/models/models/base.py (hunk 3 above)
Comment (JPXKQX) [reply models/src/anemoi/models/models/base.py:66]: Sorry for the late reply. I'm getting back to this now. I think it would be a good idea to merge this soon. 
- The order is currently obtained from the '_graph_name_hidden', so the idea is to keep it the same instead of obtaining it from the order of the 'graph.nodes' builders.
- Regarding the `additional_nodes`, this option would need to be specified in all configurations as it will be part of the Pydantic schemas, even if it is `additional_nodes: []`. In that sense, if we have to specify a key in all cases, it makes more sense to use the explicit value of 'hidden_name', as this is what we will use in the code.

Comment (JPXKQX) [reply models/src/anemoi/models/models/base.py:66]: @icedoom888 what do you think?

Comment (icedoom888) [reply models/src/anemoi/models/models/base.py:66]: Hey @JPXKQX sorry but this is still not convincing me.
I still think having the graph hidden names in the model config it's extremely confusing, regardless. 
It completely breaks the concept of swapping graph configurations effortlessly.
Eg. i send you a graph config to try out, you need to know that you also need to update the related section in the model config. Very error prone.

If you don't want to create the list dinamically as I suggested before, you should at least move the `model_config.model.model.hidden_nodes_name` in the graph configurations instead.

File: models/src/anemoi/models/layers/graph.py [hunk 7]
```diff
@@ -68,23 +68,29 @@ class NamedNodesAttributes(nn.Module):
     attr_ndims: dict[str, int]
     trainable_tensors: dict[str, TrainableTensor]
 
-    def __init__(self, num_trainable_params: int, graph_data: HeteroData) -> None:
+    def __init__(self, num_trainable_params: dict[str, int], graph_data: HeteroData) -> None:
```
Comment (MeraX) [inline models/src/anemoi/models/layers/graph.py]: I would recommend aligning the `num_trainable_params` variable name with the config key `trainable_parameters`. (here and in `define_fixed_attributes´)

Comment (JPXKQX) [reply models/src/anemoi/models/layers/graph.py]: I think that makes sense. What would you prefer? We could use num_trainable_parameters in the config or trainable_params here

File: models/src/anemoi/models/models/base.py (hunk 3 above)
Comment (JPXKQX) [reply models/src/anemoi/models/models/base.py:66]: I think that's not the case, you can still swap the graphs as before. 

This option is not a requirement for creating the graph, but rather an option to define how to use the graph in the model. When creating a graph with the graph config file, these options are not used. They are only used when creating the model.

Comment (JPXKQX) [reply models/src/anemoi/models/models/base.py:66]: If we specify the list dynamically, we still need this extra argument in the model class, `additional_nodes`. It would have the same "problem" you describe now. To me, there are 2 independent decisions here:
- whether this configuration should be in `graph` or in `model`. Following also issue #30, it is clear to me that this option belongs to the model.
- the other question is whether it is more appropriate to have a `hidden_nodes_names` argument or an `additional_nodes`, which I also think makes more sense to use the `hidden_nodes_names` as the additional_nodes will have to be specified in the configs (even if they are null because of the schemas).

Comment (JPXKQX) [reply models/src/anemoi/models/models/base.py:66]: @icedoom888

File: models/src/anemoi/models/layers/graph.py (hunk 7 above)
Comment (MeraX) [reply models/src/anemoi/models/layers/graph.py]: I would suggest either `num_trainable_parameters` , as long as it is a dict of integers. Otherwise, I would adjust the code to the config and use `trainable_parameters`. This also produces less changes in the configs.

Comment (JPXKQX) [reply models/src/anemoi/models/layers/graph.py]: Great, thanks!

File: training/src/anemoi/training/train/tasks/base.py
Comment (icedoom888) [inline training/src/anemoi/training/train/tasks/base.py]: Change docstring according to new changes

Comment (icedoom888) [inline training/src/anemoi/training/train/tasks/base.py]:
```suggestion
    graph_data : HeteroData
```

File: models/src/anemoi/models/models/hierarchical_autoencoder.py
Comment (ssmmnn11) [inline models/src/anemoi/models/models/hierarchical_autoencoder.py:27]: bypasses BaseGraphModel.__init__, will that then break?

File: training/src/anemoi/training/diagnostics/callbacks/plot.py [hunk 8]
```diff
@@ -471,8 +471,9 @@ def _plot(
             for name in self.parameters
         }
         if self.latlons is None:
```
Comment (ssmmnn11) [inline training/src/anemoi/training/diagnostics/callbacks/plot.py:473]: if self.latlons is None this should break here.

File: training/src/anemoi/training/diagnostics/callbacks/plot.py
Comment (ssmmnn11) [inline training/src/anemoi/training/diagnostics/callbacks/plot.py:583]: multi dataset ready?

Comment (ssmmnn11) [inline training/src/anemoi/training/diagnostics/callbacks/plot.py:657]: single dataset?

File: training/src/anemoi/training/diagnostics/callbacks/plot.py (hunk 8 above)
Comment (JPXKQX) [reply training/src/anemoi/training/diagnostics/callbacks/plot.py:473]: I haven't tested this, but it wasn't working before because `pl_module.model.model.graph_data` is already a dictionary of graphs in main

File: training/src/anemoi/training/diagnostics/callbacks/plot.py
Comment (JPXKQX) [reply training/src/anemoi/training/diagnostics/callbacks/plot.py:583]: I don't know, same as before

**Review by ssmmnn11 — APPROVED** [review]
ok!
